plugins/defaultplugins/l2plugin/xconnect_config.go

```diff
@@ -16,330 +16,297 @@ package l2plugin
-		if err != nil {
-			plugin.Log.Errorf("Setting l2xConnect failed: %v", err)
+		plugin.log.Debugf("Removing obsolete l2xConnect %s-%s", oldXc.ReceiveInterface, oldXc.TransmitInterface)
+		if err := vppcalls.DeleteL2XConnect(rxIfIdx, oldTxIfIdx, plugin.vppChan, plugin.stopwatch); err != nil {
+			plugin.log.Errorf("Deleted obsolete l2xConnect failed: %v", err)
 			return err
 		}
+		plugin.xcIndexes.UnregisterName(oldXc.ReceiveInterface)
+		return nil
 	}
-
-	meta := XConnectMeta{
-		TransmitInterface: newConfig.TransmitInterface,
+	// Replace existing entry
+	if err := vppcalls.AddL2XConnect(rxIfIdx, txIfIdx, plugin.vppChan, plugin.stopwatch); err != nil {
+		plugin.log.Errorf("Replacing l2xConnect failed: %v", err)
+		return err
 	}
-	plugin.XcIndexes.RegisterName(newConfig.ReceiveInterface, rxIfaceIdx, &meta)
-	plugin.XcIndexSeq++
-
-	plugin.Log.Infof("L2 xConnect pair %v modified", newConfig.ReceiveInterface)
+	plugin.xcIndexes.RegisterName(newXc.ReceiveInterface, plugin.xcIndexSeq, newXc)
+	plugin.xcIndexSeq++
+	plugin.log.Debugf("Modifying XConnect: new entry %s-%s added", newXc.ReceiveInterface, newXc.TransmitInterface)
 
 	return nil
 }
 
-// DeleteXConnectPair processes the NB config and propagates it to bin api calls.
-func (plugin *XConnectConfigurator) DeleteXConnectPair(xConnectPairInput *l2.XConnectPairs_XConnectPair) error {
-	plugin.Log.Infof("Removing L2 xConnect pair %v", xConnectPairInput.ReceiveInterface)
-
-	if err := plugin.deleteL2XConnectPair(xConnectPairInput.ReceiveInterface, xConnectPairInput.TransmitInterface); err != nil {
+// DeleteXConnectPair removes XConnect if possible. Note: Xconnect pair cannot be removed if any interface is missing.
+func (plugin *XConnectConfigurator) DeleteXConnectPair(xc *l2.XConnectPairs_XConnectPair) error {
+	plugin.log.Infof("Removing L2 xConnect pair %s-%s", xc.ReceiveInterface, xc.TransmitInterface)
+	if err := plugin.validateConfig(xc); err != nil {
 		return err
 	}
-
-	plugin.Log.Infof("L2 xConnect pair %v removed", xConnectPairInput.ReceiveInterface)
+	// If receive interface is missing, XConnect entry is not configured on the VPP.
+	rxIfIdx, _, rxFound := plugin.ifIndexes.LookupIdx(xc.ReceiveInterface)
+	if !rxFound {
+		plugin.log.Debugf("XC Del: Receive interface %s not found.", xc.ReceiveInterface)
+		// Remove from all caches
+		plugin.xcIndexes.UnregisterName(xc.ReceiveInterface)
+		plugin.xcAddCacheIndexes.UnregisterName(xc.ReceiveInterface)
+		plugin.xcDelCacheIndexes.UnregisterName(xc.ReceiveInterface)
+		return nil
+	}
+	// Verify transmit interface. If it is missing, XConnect cannot be removed and will be put to cache for deleted
+	// interfaces
+	txIfIdx, _, txFound := plugin.ifIndexes.LookupIdx(xc.TransmitInterface)
+	if !txFound {
+		plugin.log.Debugf("XC Del: Transmit interface %s not found.", xc.ReceiveInterface)
+		plugin.putOrUpdateCache(xc, false)
+		// Remove from other caches
+		plugin.xcIndexes.UnregisterName(xc.ReceiveInterface)
```

Comment:
Same here as well and below. Also, doesn't need to be unregistered from `xcDelCacheIndexes`?

Reply:
xc.ReceiveInterface is correct, that's the xconnect id

xcDelCacheIndexes is not cleared intentionally. On the VPP, if you remove rx interface, the whole xconnect entry is removed by vpp (and removed from all caches in configurator). But if you remove tx interface, xconnect stays on the VPP with some strange state, it looks like this:

vpp# sh mode
l3 local0  
l2 xconnect tapcli-0 DELETED

and such a xconnect cannot be removed, only replaced. So it is added only to 'del' cache so configurator knows about it.